### ISSUE TITLE
[Feature] Inference Carbon Tracking — Per-Request Emission Measurement and Serving Framework Integration

### ISSUE THREAD

**Issue by Tamish03** [mlco2/codecarbon#1178]
Summary
CodeCarbon is designed around training loops. In production, inference is often where the majority of a model's lifetime carbon is spent — a model fine-tuned once may serve millions of requests over years. This proposal adds a first-class InferenceCarbonTracker that measures per-request emissions, tracks batching efficiency, and integrates with common serving frameworks.

Problem
Consider a production deployment:

BERT fine-tuning (one-time): 0.3 kg CO2
BERT serving at 1,000 requests/day for 2 years: estimated 18–40 kg CO2 (varies by hardware, batching, utilization)

Currently, CodeCarbon can wrap a serving loop with EmissionsTracker but:

There is no per-request granularity
There is no batch efficiency metric (carbon per sample vs batch size)
There is no latency-carbon tradeoff analysis
There are no ready-made integrations for FastAPI, Flask, or TorchServe
The overhead model is designed for long-running loops, not short-duration API calls

This means organizations reporting AI carbon footprints are systematically underreporting — they capture training but miss serving.

Proposed Solution
A dedicated InferenceCarbonTracker with:

Per-request tracking via context manager or decorator — minimal overhead (~0.5ms)
Session aggregation — total carbon across a serving session
Batch efficiency analysis — carbon per sample as a function of batch size
Framework middleware — FastAPI and Flask integrations

API Design
pythonfrom codecarbon.inference import InferenceCarbonTracker

# Session-level tracking
tracker = InferenceCarbonTracker(
    model_name="bert-sentiment-v2",
    hardware="T4",
    region="us-east-1"
)
tracker.start_session()

# Per-request tracking
with tracker.track_request(batch_size=8):
    outputs = model(inputs)

# Get session summary
session = tracker.end_session()
print(session.total_requests)          # 1,247
print(session.total_emissions_kg)      # 0.0034
print(session.avg_emissions_per_req)   # 2.7e-6 kg CO2 / request
print(session.carbon_per_sample_curve) # dict: {1: x, 4: y, 8: z, 16: w}

# Decorator usage
@tracker.track_request
def predict(batch):
    return model(batch)

# FastAPI middleware
app.add_middleware(
    InferenceCarbonMiddleware,
    tracker=tracker,
    add_response_headers=True  # adds X-Carbon-Per-Request header
)

Architecture
codecarbon/inference/
    ├── __init__.py
    ├── tracker.py               # InferenceCarbonTracker
    ├── request_context.py       # RequestContext dataclass
    ├── session.py               # InferenceSession aggregation
    ├── batch_analyzer.py        # Carbon vs batch size analysis
    └── adapters/
          ├── fastapi_middleware.py
          ├── flask_decorator.py
          └── torchserve_handler.py

Per-Request Overhead Strategy
Standard EmissionsTracker has ~100ms startup overhead (hardware detection, geolocation). This is fine for training but unacceptable per inference request.
Proposed approach: InferenceCarbonTracker.start_session() runs the expensive initialization once. Per-request tracking uses a lightweight power sampling approach: instead of re-querying hardware on every request, it:

Maintains a running power estimate from a background thread (every 500ms)
Multiplies power × request_duration for each request
Accumulates into session totals

This gives ~0.5ms overhead per request with <5% accuracy tradeoff vs continuous sampling.

Open Questions for Maintainers

Should per-request carbon be written to a request log CSV, or only aggregated to session-level?
Is adding X-Carbon-Per-Request response headers in the FastAPI middleware too opinionated, or useful enough to include?
Should InferenceCarbonTracker reuse EmissionsTracker internally (composition) or be a parallel implementation?
What's the project's policy on framework integrations that introduce optional dependencies (fastapi, flask)?

Implementation Plan

 PR 1: InferenceCarbonTracker core + InferenceSession dataclass (no framework deps)
 PR 2: BatchEfficiencyAnalyzer + carbon-per-sample-vs-batch-size analysis
 PR 3: FastAPI middleware (optional dep, guarded import)
 PR 4: Flask decorator (optional dep)
 PR 5: Dashboard view for inference carbon history
